Source/ExecutionEngine/ImplementationTask.cs

```diff
@@ -62,40 +72,59 @@ public class ImplementationTask : IImplementationTask {
   }
 
   private CancellationTokenSource? cancellationSource;
+  private ReplaySubject<IVerificationStatus>? status;
 
   public void Cancel() {
     cancellationSource?.Cancel();
-    cancellationSource = null;
   }
 
+  public bool IsIdle => cancellationSource == null;
+
   public IObservable<IVerificationStatus>? TryRun()
   {
     if (CacheStatus is Completed) {
       return null;
     }
 
-    var alreadyRunning = cancellationSource != null;
-    if (alreadyRunning) {
-      return null;
+    lock (myLock) {
```

Comment:
Shouldn't the lock also contain the previous if statement?
Two threads could arrive at this lock at the same time and a third one sets `CacheStatus` as `Completed` before it enters the lock.
Then there could be race conditions.